File: src/main/java/ru/org/linux/user/RegisterController.java
/*
 * Copyright 1998-2012 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.user;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.web.authentication.RememberMeServices;
import org.springframework.stereotype.Controller;
import org.springframework.validation.Errors;
import org.springframework.web.bind.WebDataBinder;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.ModelAndView;
import org.springframework.web.servlet.view.RedirectView;
import ru.org.linux.auth.*;
import ru.org.linux.site.Template;
import ru.org.linux.spring.Configuration;
import ru.org.linux.util.EmailService;
import ru.org.linux.util.ExceptionBindingErrorProcessor;
import ru.org.linux.util.LorHttpUtils;

import javax.mail.internet.InternetAddress;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpSession;
import javax.validation.Valid;

@SuppressWarnings("ProhibitedExceptionDeclared")
@Controller
public class RegisterController {
  private static final Log logger = LogFactory.getLog(RegisterController.class);

  private CaptchaService captcha;
  private IPBlockDao ipBlockDao;

  @Autowired
  RememberMeServices rememberMeServices;

  @Autowired
  @Qualifier("authenticationManager")
  private AuthenticationManager authenticationManager;

  @Autowired
  private UserDetailsServiceImpl userDetailsService;


  @Autowired
  private UserDao userDao;

  @Autowired
  private EmailService emailService;

  @Autowired
  private Configuration configuration;

  @Autowired
  private PasswordVerify passwordVerify;

  @Autowired
  public void setCaptcha(CaptchaService captcha) {
    this.captcha = captcha;
  }

  @Autowired
  public void setIpBlockDao(IPBlockDao ipBlockDao) {
    this.ipBlockDao = ipBlockDao;
  }

  @RequestMapping(value = "/register.jsp", method = RequestMethod.GET)
  public ModelAndView register(
    @ModelAttribute("form") RegisterRequest form,
    HttpServletResponse response
  ) {
      response.setHeader("Cache-Control", "no-store, no-cache, must-revalidate");
      return new ModelAndView("register");
  }

  @RequestMapping(value = "/register.jsp", method = RequestMethod.POST)
  public ModelAndView doRegister(
    HttpServletRequest request,
    @Valid @ModelAttribute("form") RegisterRequest form,
    Errors errors
  ) throws Exception {
    HttpSession session = request.getSession();
    Template tmpl = Template.getTemplate(request);

    if (!errors.hasErrors()) {
      captcha.checkCaptcha(request, errors);

      if (session.getAttribute("register-visited") == null) {
        logger.info("Flood protection (not visited register.jsp) " + request.getRemoteAddr());
        errors.reject(null, "Временная ошибка, попробуйте еще раз");
      }

      ipBlockDao.checkBlockIP(request.getRemoteAddr(), errors, tmpl.getCurrentUser());

      if (userDao.isUserExists(form.getNick())) {
        errors.rejectValue("nick", null, "пользователь " + form.getNick() + " уже существует");
      }

      if (userDao.getByEmail(new InternetAddress(form.getEmail()).getAddress().toLowerCase(), false) != null) {
        errors.rejectValue("email", null, "пользователь с таким e-mail уже зарегистрирован. " +
                "Если вы забыли параметры своего аккаунта, воспользуйтесь формой восстановления пароля");
      }
    }

    if (!errors.hasErrors()) {
      InternetAddress mail = new InternetAddress(form.getEmail().toLowerCase());

      int userid = userDao.createUser("", form.getNick(), form.getPassword(), "", mail, "", "");

      String logmessage = "Зарегистрирован пользователь " + form.getNick() + " (id=" + userid + ") " + LorHttpUtils.getRequestIP(request);
      logger.info(logmessage);

      emailService.sendEmail(form.getNick(), mail.getAddress(), true);
    } else {
      return new ModelAndView("register");
    }

    return new ModelAndView("action-done", "message", "Добавление пользователя прошло успешно. Ожидайте письма с кодом активации.");
  }


  @RequestMapping(value="/activate.jsp", method= RequestMethod.GET)
  public ModelAndView activateForm() {
    return new ModelAndView("activate");
  }

  @RequestMapping(value = "/activate.jsp", method = RequestMethod.POST, params = "action")
  public ModelAndView activateNew(
    HttpServletRequest request,
    HttpServletResponse response,
    @RequestParam String activation,
    @RequestParam String nick,
    @RequestParam String passwd
  ) throws Exception {
    UsernamePasswordAuthenticationToken token = new UsernamePasswordAuthenticationToken(nick, passwd);
    try {
      UserDetailsImpl details = (UserDetailsImpl) userDetailsService.loadUserByUsername(nick);
      if(details.getUser().isActivated()) {
        return new ModelAndView(new RedirectView("/"));
      }
      token.setDetails(details);
      Authentication auth = authenticationManager.authenticate(token);
      UserDetailsImpl userDetails = (UserDetailsImpl)auth.getDetails();
      String regcode = userDetails.getUser().getActivationCode(configuration.getSecret());
      if(regcode.equals(activation)) {
        userDao.activateUser(userDetails.getUser());
        details = (UserDetailsImpl) userDetailsService.loadUserByUsername(nick);
        token.setDetails(details);
        auth = authenticationManager.authenticate(token);
      } else {
        throw new AccessViolationException("Bad activation code");
      }
      SecurityContextHolder.getContext().setAuthentication(auth);
      rememberMeServices.loginSuccess(request, response, auth);
      AuthUtil.updateLastLogin(auth, userDao);
    } catch (Exception e) {
      throw new AccessViolationException(e.getMessage());
    }
    return new ModelAndView(new RedirectView("/"));
  }


  @RequestMapping(value = "/activate.jsp", method = RequestMethod.POST, params = "!action")
  public ModelAndView activate(
    HttpServletRequest request,
    @RequestParam String activation
  ) throws Exception {
    Template tmpl = Template.getTemplate(request);

    if (!tmpl.isSessionAuthorized()) {
      throw new AccessViolationException("Not authorized!");
    }

    User user = tmpl.getCurrentUser();

    String newEmail = userDao.getNewEmail(user);

    if (newEmail == null) {
      throw new AccessViolationException("new_email == null?!");
    }

    String regcode = user.getActivationCode(configuration.getSecret(), newEmail);

    if (!regcode.equals(activation)) {
      throw new AccessViolationException("Bad activation code");
    }

    userDao.acceptNewEmail(user);

    return new ModelAndView(new RedirectView("/people/" + user.getNick() + "/profile"));
  }

  @InitBinder("form")
  public void requestValidator(WebDataBinder binder) {
    binder.setValidator(new RegisterRequestValidator(passwordVerify));
    binder.setBindingErrorProcessor(new ExceptionBindingErrorProcessor());
  }
}

File: src/main/scala/ru/org/linux/user/RegisterController.scala
/*
 * Copyright 1998-2023 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package ru.org.linux.user

import com.typesafe.scalalogging.StrictLogging
import io.circe.Json
import io.circe.syntax.EncoderOps
import org.jasypt.util.text.AES256TextEncryptor
import org.joda.time.DateTime
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.core.io.ResourceLoader
import org.springframework.security.authentication.{AuthenticationManager, BadCredentialsException, UsernamePasswordAuthenticationToken}
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.web.authentication.RememberMeServices
import org.springframework.stereotype.Controller
import org.springframework.validation.Errors
import org.springframework.web.bind.WebDataBinder
import org.springframework.web.bind.annotation.*
import org.springframework.web.servlet.ModelAndView
import org.springframework.web.servlet.view.RedirectView
import ru.org.linux.auth.AuthUtil.AuthorizedOnly
import ru.org.linux.auth.*
import ru.org.linux.email.EmailService
import ru.org.linux.spring.SiteConfig
import ru.org.linux.util.{ExceptionBindingErrorProcessor, LorHttpUtils, StringUtil}

import javax.mail.internet.InternetAddress
import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
import javax.validation.Valid
import scala.jdk.CollectionConverters.*

@Controller
class RegisterController(captcha: CaptchaService, rememberMeServices: RememberMeServices,
                         @Qualifier("authenticationManager") authenticationManager: AuthenticationManager,
                         userDetailsService: UserDetailsServiceImpl, userDao: UserDao, emailService: EmailService,
                         siteConfig: SiteConfig, userService: UserService, invitesDao: UserInvitesDao,
                         resourceLoader: ResourceLoader) extends StrictLogging {
  private val registerRequestValidator = new RegisterRequestValidator(resourceLoader)

  @RequestMapping(value = Array("/register.jsp"), method = Array(RequestMethod.GET))
  def register(@ModelAttribute("form") form: RegisterRequest, response: HttpServletResponse,
               request: HttpServletRequest, @RequestParam(required = false) invite: String): ModelAndView = {
    response.setHeader("Cache-Control", "no-store, no-cache, must-revalidate")

    if (invite!=null) {
      val emailOpt = invitesDao.emailFromValidInvite(invite)

      emailOpt match {
        case None =>
          throw new AccessViolationException("Код приглашения не действителен")
        case Some(email) =>
          form.setEmail(email)
      }
      new ModelAndView("register", "invite", invite)
    } else {
      if (userService.canRegister(request.getRemoteAddr)) {
        new ModelAndView("register", "permit", makePermit)
      } else {
        new ModelAndView("no-register")
      }
    }
  }

  private def makePermit: String = {
    val key = siteConfig.getSecret
    val message = s"permit:${DateTime.now().plusHours(1).getMillis}"

    val textEncryptor = new AES256TextEncryptor
    textEncryptor.setPassword(key)
    textEncryptor.encrypt(message)
  }

  private def checkPermit(permit: String): Boolean = {
    val key = siteConfig.getSecret
    val textEncryptor = new AES256TextEncryptor
    textEncryptor.setPassword(key)

    textEncryptor.decrypt(permit).split(":", 2) match {
      case Array("permit", date) =>
        val decodedDate = new DateTime(date.toLong)
        logger.debug(s"Decoded permit date: $decodedDate")
        decodedDate.isAfterNow
      case other =>
        logger.warn(s"Invalid permit - decrypted: $other")
        false
    }
  }

  @RequestMapping(value = Array("/register.jsp"), method = Array(RequestMethod.POST))
  def doRegister(request: HttpServletRequest, @Valid @ModelAttribute("form") form: RegisterRequest,
                 errors: Errors, @RequestParam(required = false) invite: String,
                 @RequestParam(required = false) permit: String): ModelAndView = {
    if (invite==null && permit == null) {
      return new ModelAndView("no-register")
    } else {
      if (invite!=null) {
        val emailOpt = invitesDao.emailFromValidInvite(invite)

        emailOpt match {
          case None =>
            throw new AccessViolationException("Код приглашения не действителен")
          case Some(email) =>
            form.setEmail(email)
        }
      } else if (!checkPermit(permit)) {
        return new ModelAndView("no-register")
      }
    }

    if (!errors.hasErrors) {
      if (invite==null) {
        captcha.checkCaptcha(request, errors)
      }

      if (userDao.isUserExists(form.getNick) || userDao.hasSimilarUsers(form.getNick)) {
        errors.rejectValue("nick", null, "Это имя пользователя уже используется. Пожалуйста выберите другое имя.")
      }

      if (userDao.getByEmail(new InternetAddress(form.getEmail).getAddress.toLowerCase, false) != null) {
        errors.rejectValue("email", null, "пользователь с таким e-mail уже зарегистрирован. " +
          "Если вы забыли параметры своего аккаунта, воспользуйтесь формой восстановления пароля.")
      }
    }

    if (!errors.hasErrors) {
      val mail = new InternetAddress(form.getEmail.toLowerCase)
      val userid = userService.createUser(nick = form.getNick, password = form.getPassword, mail = mail,
        ip = request.getRemoteAddr, invite = Option(invite), userAgent = Option(request.getHeader("user-agent")),
        language = Option(request.getHeader("accept-language")))

      logger.info(s"Зарегистрирован пользователь ${form.getNick} (id=$userid) ${LorHttpUtils.getRequestIP(request)}")

      emailService.sendRegistrationEmail(form.getNick, mail.getAddress, isNew = true)

      new ModelAndView(
        "action-done",
        "message",
        "Добавление пользователя прошло успешно. Ожидайте письма с кодом активации.")
    } else {
      val params = Map("invite" -> invite, "permit" -> permit)
      new ModelAndView("register", params.asJava)
    }
  }

  private def activationFormParams(nick: String, activation: String) = {
    val nickSanitized = Option(nick).filter(StringUtil.checkLoginName).orNull
    val activationSanitized = Option(activation).filter(_.forall(_.isLetterOrDigit)).orNull

    Map(
      "nick" -> nickSanitized,
      "activation" -> activationSanitized
    )
  }

  @RequestMapping(value = Array("/activate", "/activate.jsp"), method = Array(RequestMethod.GET))
  def activateForm(@RequestParam(required = false) nick: String,
                   @RequestParam(required = false) activation: String): ModelAndView = {
    new ModelAndView("activate", activationFormParams(nick, activation).asJava)
  }

  @RequestMapping(value = Array("/activate", "/activate.jsp"), method = Array(RequestMethod.POST), params = Array("action"))
  def activateNew(request: HttpServletRequest, response: HttpServletResponse,
                  @RequestParam activation: String, @RequestParam nick: String,
                  @RequestParam passwd: String): ModelAndView = {
    try {
      val details = userDetailsService.loadUserByUsername(nick).asInstanceOf[UserDetailsImpl]

      if (!details.getUser.isActivated) {
        val token = new UsernamePasswordAuthenticationToken(nick, passwd)

        token.setDetails(details)

        val auth = authenticationManager.authenticate(token)
        val userDetails = auth.getDetails.asInstanceOf[UserDetailsImpl]
        val regcode = userDetails.getUser.getActivationCode(siteConfig.getSecret)

        if (regcode.equalsIgnoreCase(activation)) {
          userDao.activateUser(userDetails.getUser)

          val updatedDetails = userDetailsService.loadUserByUsername(nick).asInstanceOf[UserDetailsImpl]
          token.setDetails(updatedDetails)
          val updatedAuth = authenticationManager.authenticate(token)

          SecurityContextHolder.getContext.setAuthentication(updatedAuth)
          rememberMeServices.loginSuccess(request, response, updatedAuth)
          AuthUtil.updateLastLogin(updatedAuth, userDao)

          new ModelAndView(new RedirectView("/"))
        } else {
          val params = activationFormParams(nick, activation) + ("error" -> "Неправильный код активации")
          new ModelAndView("activate", params.asJava)
        }
      } else {
        new ModelAndView(new RedirectView("/"))
      }
    } catch {
      case _: UsernameNotFoundException =>
        val params = activationFormParams(nick, activation) + ("error" -> "Пользователь не найден")
        new ModelAndView("activate", params.asJava)
      case _: BadCredentialsException =>
        val params = activationFormParams(nick, activation) + ("error" -> "Неправильный логин или пароль")
        new ModelAndView("activate", params.asJava)
    }
  }

  @RequestMapping(value = Array("/activate", "/activate.jsp"), method = Array(RequestMethod.POST), params = Array("!action"))
  def activate(@RequestParam activation: String): ModelAndView = AuthorizedOnly { currentUser =>
    val newEmail = userDao.getNewEmail(currentUser.user)

    if (newEmail == null) {
      throw new AccessViolationException("new_email == null?!")
    }

    val regcode = currentUser.user.getActivationCode(siteConfig.getSecret, newEmail)

    if (!regcode.equalsIgnoreCase(activation)) {
      val params = activationFormParams(currentUser.user.getNick, activation) + ("error" -> "Неправильный код активации")

      new ModelAndView("activate", params.asJava)
    } else {
      userDao.acceptNewEmail(currentUser.user, newEmail)

      new ModelAndView(new RedirectView("/people/" + currentUser.user.getNick + "/profile"))
    }
  }

  @ResponseBody
  @RequestMapping(path = Array("check-login"))
  def ajaxLoginCheck(@RequestParam nick: String): Json = {
    (if (nick.isEmpty) {
      "Не задан nick."
    } else if (!StringUtil.checkLoginName(nick)) {
      "Некорректное имя пользователя."
    } else if (nick != null && nick.length > User.MAX_NICK_LENGTH) {
      "Слишком длинное имя пользователя."
    } else if (userDao.isUserExists(nick) || userDao.hasSimilarUsers(nick)) {
      "Это имя пользователя уже используется. Пожалуйста выберите другое имя."
    } else {
      "true"
    }).asJson
  }

  @InitBinder(Array("form"))
  def requestValidator(binder: WebDataBinder):Unit = {
    binder.setValidator(registerRequestValidator)
    binder.setBindingErrorProcessor(new ExceptionBindingErrorProcessor)
  }

  @RequestMapping(value = Array("create-invite"), method = Array(RequestMethod.GET))
  def createInviteForm(): ModelAndView = AuthorizedOnly { currentUser =>
    if (!userService.canInvite(currentUser.user)) {
      throw new AccessViolationException("Вы не можете пригласить нового пользователя")
    }

    new ModelAndView("create-invite")
  }

  @RequestMapping(value = Array("create-invite"), method = Array(RequestMethod.POST))
  def createInvite(@RequestParam email: String): ModelAndView = AuthorizedOnly { currentUser =>
    if (!userService.canInvite(currentUser.user)) {
      throw new AccessViolationException("Вы не можете пригласить нового пользователя")
    }

    if (userDao.getByEmail(email, false) != null) {
      throw new AccessViolationException("Пользователь с этим адресом уже зарегистрирован")
    }

    val parsedEmail = new InternetAddress(email)

    if (!registerRequestValidator.isGoodDomainEmail(parsedEmail)) {
      throw new AccessViolationException("Некорректный email домен")
    }

    val (token, validUntil) = invitesDao.createInvite(currentUser.user, email)

    emailService.sendInviteEmail(currentUser.user, email, token, validUntil)

    new ModelAndView("action-done", "message", s"Приглашение отправлено по адресу $email")
  }
}